Proposal: Opening/Closing Mechanism for Zip Files

Michael5601 opened the pull request:
# Proposal: Opening/Closing Mechanism for Zip Files

## Background
The Eclipse IDE has no built in functionality to open Zip Files and read or manipulate their content. Because of this, other operations like searching inside of Zip Files or comparing two Zip Files were also not possible.

## Description
This PR introduces UI support for accesing the opening/closing mechanism for Zip Files over the UI-menu. It is accessed by right-clicking the Zip File that should be opened and then clicking on "Open Zip File". Closing is accessed the same way but when right-clicking an opened Zip File.

Please see [Plaform PR](https://github.com/eclipse-platform/eclipse.platform/pull/1413) for the PR on the repository **eclipse.platform** for the platform implementation and further information.

Co-Authored-By: CodeLtDave [david.erdoes@vector.com](mailto:david.erdoes@vector.com)

iloveeclipse reviewed (changes requested):
> Co-Authored-By: CodeLtDave [david.erdoes@vector.com](mailto:david.erdoes@vector.com)

Here and on other PR: **ALL** contributors must sign ECA. 

If `david.erdoes@vector.com` has not signed ECA, we can't  merge this PR.

jukzi commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/fileSystem/zip/CloseZipFileHandler.java:
```diff
@@ -0,0 +1,59 @@
+		try {
+			ZipFileTransformer.closeZipFile((IFolder) element);
+		} catch (Exception e) {
+			MessageDialog.openError(shell, "Error", "Error opening zip file"); //$NON-NLS-1$ //$NON-NLS-2$
```
log this error instead of error dialog/std.err

Michael5601 replied:
thanks for the input. I commited the changed under 297459a371fb37bc1c023abc6ec0da099d6688cc. But I wonder if the user should get a response that closing the zip file did not work. This was the reason behind the MessageDialog even though the message is wrong and should be "Error closing zip file".

jukzi commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/fileSystem/zip/OpenZipFileHandler.java:
```diff
@@ -0,0 +1,79 @@
+						monitor.beginTask("Opening Zip File", 5); //$NON-NLS-1$
+						try {
+							ZipFileTransformer.openZipFile((IFile) element, monitor, true);
+						} catch (URISyntaxException | CoreException e) {
```
The exception stacktrace is lost here. Make sure full exception can somehow be seen in log

Michael5601 replied:
In 21ca9345a09d9a8bdcad30bd5712051ed5172a25 I added a log before the MessageDialog to preserve the full exception.

jukzi commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/fileSystem/zip/OpenZipFileHandler.java:
```diff
@@ -0,0 +1,79 @@
+					}
+				});
+			} catch (InvocationTargetException e) {
+				e.printStackTrace();
```
do not use std.err

Michael5601 replied:
In this 297459a371fb37bc1c023abc6ec0da099d6688cc I removed the usage of std.err.

jukzi commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/fileSystem/zip/OpenZipFileHandler.java:
```diff
@@ -0,0 +1,79 @@
+				});
+			} catch (InvocationTargetException e) {
+				e.printStackTrace();
+			} catch (InterruptedException e) {
```
typically InterruptedException is handled silently

Michael5601 replied:
This is the same as in the [comment](https://github.com/eclipse-platform/eclipse.platform.ui/pull/1947#discussion_r1633350494) above. I want the user to be informed about the failed opening of the zip file.

jukzi replied:
That makes more sense in the Open Action then during close, but please do not misuse InterruptedException for that

Michael5601 replied:
Thank you for your input last week. I removed the ProgressDialog in 297459a371fb37bc1c023abc6ec0da099d6688cc and thus no InterruptedException is thrown anymore.

jukzi commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/fileSystem/zip/ZipFileSystemContributor.java:
```diff
@@ -0,0 +1,84 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
```
2024 Vector?

Michael5601 replied:
This method is copied unchanged from the example project. We thought this would be the best place for it but did not change it, thus the Copyright is still the old one.

jukzi commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/fileSystem/zip/ZipFileSystemContributor.java:
```diff
@@ -0,0 +1,84 @@
+		}
+	}
+
+	/* (non-Javadoc)
```
remove this javadoc. javadoc is automatically inherited for overriden methods

Michael5601 replied:
You are right. I removed the javadoc.

jukzi commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/fileSystem/zip/ZipFileSystemContributor.java:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	@Override
+	public URI getURI(String pathString) {
```
this method feels confusing complicated. please give an example in java comment what the resulting URI will look like.

jukzi commented:
Please post some screenshots about the workflow.

Commit:
Proposal: Opening/Closing Mechanism for Zip Files

Proposal: Opening/Closing Mechanism for Zip Files

The Eclipse IDE has no built in functionality to open Zip Files and read or manipulate their content. Because of this, other operations like searching inside of Zip Files or comparing two Zip Files were also not possible.

This PR introduces UI support for accesing the opening/closing mechanism for Zip Files over the UI-menu. It is accessed by right-clicking the Zip File that should be opened and then clicking on "Open Zip File". Closing is accessed the same way but when right-clicking an opened Zip File.

Please see #1408 for the PR on the repository **eclipse.platform** for the platform implementation and further information.

Co-Authored-By: David Erdös <David.Erdoes@vector.com>

Michael5601 commented:
I want to clarify the workflow with the following screenshots:

At first the Zip File is closed and it is not possible to view/modify or search its contents. 
![ClosedZip](https://github.com/eclipse-platform/eclipse.platform.ui/assets/102024826/dbcd8c5e-fb56-4918-ae06-301d2776ab0a)

By right-clicking and selection "Open Zip File" the Zip File can be opened.
![Menu](https://github.com/eclipse-platform/eclipse.platform.ui/assets/102024826/122ea482-740c-48d5-af78-a4f2a3bc9732)

Opening the Zip File allows using it like a normal folder with all operations supported.
![OpenedZip](https://github.com/eclipse-platform/eclipse.platform.ui/assets/102024826/45cc9448-e2b9-4d01-b3b4-db41898e3202)

Opening a resource inside of the zip opens the editor for the resource.
![OpenedResource](https://github.com/eclipse-platform/eclipse.platform.ui/assets/102024826/62cba319-a683-4277-a618-e73b13a59eee)

Opened Zip Files can be closed by right-clicking and selecting "Close Zip File".
![Menu2](https://github.com/eclipse-platform/eclipse.platform.ui/assets/102024826/5bf3de11-925b-4e8f-9b80-91fa5dc987d9)

After closing the zip file we have the same state as in the beginning.
![ClosedZip](https://github.com/eclipse-platform/eclipse.platform.ui/assets/102024826/5f20dc01-ee1c-4568-931d-1f3b3cb852ff)

For other archive types than "zip" the button will not show in the menu when right-clicking the archive.
I hope these screenshots could help to understand the workflow.

jukzi commented:
> I hope these screenshots could help to understand the workflow.

Makes sense. I would have expected open/close to be more at the top, as the normal "Open" action is:
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/51790620/46f6f9a9-2748-494e-8cb9-d0952b8c4389)

jukzi commented:
Just another idea: explicit open/closing zipfiles could be avoided by adding a node "Archives" like JDT already does for "Referenced Libraries" which makes it possible to explore the zip files contents too.
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/51790620/d8ad5a18-7a82-4560-a07d-7ec87fffd67f)

Commit:
change Open/Close Zip error handling

This change removes the ProgressDialog in the OpenZipHandler because the operation is so fast that it is not needed.
Errors are logged instead of being printed.

Commit:
log exception so that stack trace does not get lost

Michael5601 commented:
> Just another idea: explicit open/closing zipfiles could be avoided by adding a node "Archives" like JDT already does for "Referenced Libraries" which makes it possible to explore the zip files contents too. ![image](https://private-user-images.githubusercontent.com/51790620/339724298-d8ad5a18-7a82-4560-a07d-7ec87fffd67f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MTMyMDksIm5iZiI6MTcxODcxMjkwOSwicGF0aCI6Ii81MTc5MDYyMC8zMzk3MjQyOTgtZDhhZDVhMTgtN2E4Mi00NTYwLWEwN2QtN2VjODdmZmZkNjdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE4VDEyMTUwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljMGYyMjY2OWMwZDJiZGNhYzVkMjVlMjE1Y2QyMDQxMmNlMzVkYzQxNzU5ZTZjZjY5MGE2YmYwMTU0MzEzMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.C4Uz_-X1g6flSYsnPwm9hZfVCU2qVrT258yjMePW-mc)

This was our initial approach to the topic, though I don't recall the exact reasons we decided against it, given the time that has passed. The main issue, I believe, was that this window relies on JDT, while we wanted to exclusively use platform content. Additionally, the search functionality only targets content within the local workspace. Since our project aimed to improve the file search at first, we quickly abandoned the JDT approach. If you know a specific way how this could work with JDT we can also examine it further.

Commit:
Failed attempt to include open behaviour to "Open" button

Please mind that the "Open Zip File" and "Close Zip File" Buttons are still implemented in the menu

Commit:
allow JAR and WAR files to be opened

Commit:
Revert "Failed attempt to include open behaviour to "Open" button"

This reverts commit 4a87a0c2a4c687fa8b4702be716becaed0118ef2.

Commit:
Implement new Decorator for zip files

CodeLtDave committed:
use jar icon for VirtualZipFolders

Commit:
remove refreshViewers

Commit:
refreshing of all viewers after opening/closing zip file

Commit:
delete added icon

Commit:
implement double click open for zip files

This implementation is not yet finished as a proper editor handling needs to be added.

Commit:
Shown In System Explorer: Workaround for proper error handling

This commit introduces as special case check when performing the Show In System Explorer Operation on files inside of a zip file. An investigation of better solutions will follow.

Commit:
adjust MessageDialog for deleting zip files

Deleting an opened zip file also deletes the resource in the local file system. An opened zip file is represented as a linked folder which shows a specific Messagedialog when being deleted indicating that only the link will be deleted. In this case the delete operation of zip files also deletes the resource in the local file system.

Please mind that one dialog exists in JDT which should also be adjusted similarily.

Commit:
additional editor close

Commit:
Revert "remove refreshViewers"

This reverts commit 6a3216ea21bcb11388e750657f89dedf3b4e0dd8.

Commit:
refactor refreshViewers

Commit:
allow close Zip File always if folder is VirtualZipFolder

Commit:
remove explicit refresh

Commit:
refresh workaround
